ompi/mca/pml/ucx/pml_ucx_datatype.c

```diff
@@ -135,20 +161,57 @@ int mca_pml_ucx_datatype_attr_del_fn(ompi_datatype_t* datatype, int keyval,
 {
     ucp_datatype_t ucp_datatype = (ucp_datatype_t)attr_val;
 
+#ifdef HAVE_UCP_REQUEST_PARAM_T
+    free((void*)datatype->pml_data);
+#else
     PML_UCX_ASSERT((uint64_t)ucp_datatype == datatype->pml_data);
-
+#endif
     ucp_dt_destroy(ucp_datatype);
     datatype->pml_data = PML_UCX_DATATYPE_INVALID;
     return OMPI_SUCCESS;
 }
 
+#ifdef HAVE_UCP_REQUEST_PARAM_T
+__opal_attribute_always_inline__ static
+pml_ucx_datatype_t *mca_pml_ucx_init_nbx_datatype(ucp_datatype_t ucp_datatype,
+                                                  size_t size,
+                                                  int is_contig_pow2)
+{
+    pml_ucx_datatype_t *pml_datatype;
+
+    pml_datatype = malloc(sizeof(*pml_datatype));
+    if (pml_datatype == PML_UCX_DATATYPE_INVALID) {
+        PML_UCX_ERROR("Failed to allocate datatype structure");
+        ompi_mpi_abort(&ompi_mpi_comm_world.comm, 1);
+    }
+
+    *pml_datatype          = pml_ucx_datatype_tmpl;
```

Comment:
i think it would be simpler that instead of defining pml_ucx_datatype_tmpl, just initialize all parts of *pml_datatype here to default values